### PR TITLE
[Feature] Loading button

### PR DESCRIPTION
**Pull request by aleksik**
## Description

This button implements the loading button by introducing `isLoading` and `loadingText` -props to the `<Button />` component.

## Demo

👉   [Storybook demo](https://city-of-helsinki.github.io/hds-demo/loading-button/iframe.html?id=components-button--loading)

### CODE REVIEW

**Comment by ronijaakkola** [comment]
Documentation is done but it is missing an interactive example of the loading button. 
image

File: packages/react/src/components/button/Button.tsx [hunk 1]
```diff
@@ -90,24 +108,39 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
       </div>
     ) : null;
 
+    const loadingSpinner = (
```
[the remark this answers is not in the document]
Comment (minevala) [reply packages/react/src/components/button/Button.tsx:111]: I'll replace

File: packages/react/src/components/button/Button.module.scss [hunk 2]
```diff
@@ -37,6 +37,29 @@
   @extend %supplementary;
 }
 
+.isLoading {
+  --background-color: transparent;
```
Comment (VilleMiekkoja) [inline packages/react/src/components/button/Button.module.scss:41]: Where are these variables applied?

File: packages/react/src/components/button/Button.tsx [hunk 3]
```diff
@@ -90,24 +108,31 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
...
         className={classNames(
           styles.button,
           styles[variant],
           styles[`theme-${theme}`],
           styles[`size-${size}`],
           fullWidth ? styles.fullWidth : '',
+          isLoading ? styles.isLoading : '',
           className,
         )}
+        onClick={isLoading ? loadingOnClick : onClick}
```
Comment (VilleMiekkoja) [inline packages/react/src/components/button/Button.tsx:130]: Was there no onClick before? onClick seems to be missing from the CommonButtonProps 🤔

File: packages/react/src/components/button/Button.tsx [hunk 4]
```diff
@@ -90,24 +108,31 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
...
 
+    const loadingOnClick: React.MouseEventHandler<HTMLButtonElement> = (event) => {
+      event.preventDefault();
+    };
+
     return (
       <button
         ref={ref}
         disabled={disabled}
+        aria-disabled={isLoading || disabled || undefined}
```
Comment (VilleMiekkoja) [inline packages/react/src/components/button/Button.tsx:119]: Should the undefined be false instead?

File: packages/react/src/components/button/__snapshots__/Button.test.tsx.snap [hunk 5]
```diff
@@ -14,3 +14,26 @@ exports[`<Button /> spec renders the component 1`] = `
...
+  <button
+    aria-disabled="true"
+    class="button primary theme-default size-default isLoading"
+    type="button"
+  >
+    <div
+      class="loadingSpinner small"
+    >
+      <div />
+      <div />
```
Comment (VilleMiekkoja) [inline packages/react/src/components/button/__snapshots__/Button.test.tsx.snap:29]: Hmm.. 3 empty divs here 🤔

File: packages/react/src/components/dateInput/__snapshots__/DateInput.test.tsx.snap [hunk 6]
```diff
@@ -2583,6 +2583,7 @@ exports[`<DateInput /> spec renders the component with default props 1`] = `
               class="hds-datepicker__bottom-buttons"
             >
               <button
+                aria-disabled="true"
```
Comment (VilleMiekkoja) [inline packages/react/src/components/dateInput/__snapshots__/DateInput.test.tsx.snap:2586]: Are these changes wanted?

File: packages/react/src/components/button/__snapshots__/Button.test.tsx.snap (hunk 5 above)
Comment (minevala) [reply packages/react/src/components/button/__snapshots__/Button.test.tsx.snap:29]: This is the html generated by LoadingSpinner component

File: packages/react/src/components/button/Button.stories.tsx [hunk 7]
```diff
@@ -75,6 +75,39 @@ export const Icons = () => (
...
+        setIsLoading(false);
+      }, 5000);
+    }
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [isLoading]);
+  return (
+    <>
+      <Button
```
[the remark this answers is not in the document]
Comment (minevala) [reply packages/react/src/components/button/Button.stories.tsx:96]: I can add one there as well

File: packages/react/src/components/dateInput/__snapshots__/DateInput.test.tsx.snap (hunk 6 above)
Comment (minevala) [reply packages/react/src/components/dateInput/__snapshots__/DateInput.test.tsx.snap:2586]: I quickly checked the datepicker code and it seems that these button were disabled

File: packages/react/src/components/button/Button.tsx (hunk 4 above)
Comment (minevala) [reply packages/react/src/components/button/Button.tsx:119]: I guess it does not render undefined properties

File: packages/react/src/components/button/Button.tsx (hunk 3 above)
Comment (minevala) [reply packages/react/src/components/button/Button.tsx:130]: It is part of the react's button-props which ButtonProps extended.

File: packages/react/src/components/button/Button.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (minevala) [reply packages/react/src/components/button/Button.stories.tsx:96]: Or I have to replace the interactive version 🤔

**Review by VilleMiekkoja — APPROVED** [review]
Nice 👍

File: packages/react/src/components/button/Button.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (minevala) [reply packages/react/src/components/button/Button.stories.tsx:102]: I added a static example too.

**Comment by ronijaakkola** [comment]
Pushed a small update to the documentation:
- Slightly altered the loading button result toast text
- Added example code of the loading button to the "React code" section

**Comment by minevala** [comment]
Why it is not supported in Core? @ronijaakkola

**Comment by ronijaakkola** [comment]
> Why it is not supported in Core? @ronijaakkola

@minevala Can it be? I suppose that changing the button + opening a toast notification would be impossible with pure HTML/CSS. All components are supported but not the example's functionality.

**Pull request merged**